azure-kusto-ingest/src/resource_manager/ingest_client_resources.rs

```diff
@@ -0,0 +1,148 @@
+fn get_resource_by_name(table: &TableV1, resource_name: String) -> Result<Vec<ResourceUri>> {
+    let storage_root_index = get_column_index(table, "StorageRoot")?;
+    let resource_type_name_index = get_column_index(table, "ResourceTypeName")?;
+
+    let resource_uris: Vec<Result<ResourceUri>> = table
+        .rows
+        .iter()
+        .filter(|r| r[resource_type_name_index] == resource_name)
+        .map(|r| {
+            ResourceUri::try_from(r[storage_root_index].as_str().ok_or(anyhow::anyhow!(
+                "Response returned from Kusto could not be parsed as a string"
+            ))?)
+        })
+        .collect();
+
+    if resource_uris.is_empty() {
+        return Err(anyhow::anyhow!(
+            "No {} resources found in the table",
+            resource_name
+        ));
+    }
+
+    resource_uris.into_iter().collect()
+}
+
+/// Helper to turn a vector of resource URIs into a vector of Azure clients of type T with the provided [ClientOptions]
+fn create_clients_vec<T>(resource_uris: &[ResourceUri], client_options: &ClientOptions) -> Vec<T>
+where
+    T: ClientFromResourceUri,
+{
+    resource_uris
+        .iter()
+        .map(|uri| T::create_client(uri.clone(), client_options.clone()))
+        .collect()
+}
+
+/// Storage of the clients required for ingestion
+#[derive(Debug, Clone)]
+pub struct InnerIngestClientResources {
+    pub secured_ready_for_aggregation_queues: Vec<QueueClient>,
+    pub temp_storage: Vec<ContainerClient>,
+}
+
+impl TryFrom<(&TableV1, &QueuedIngestClientOptions)> for InnerIngestClientResources {
+    type Error = anyhow::Error;
+
+    /// Attempts to create a new InnerIngestClientResources from the given [TableV1] and [QueuedIngestClientOptions]
+    fn try_from((table, client_options): (&TableV1, &QueuedIngestClientOptions)) -> Result<Self> {
+        let secured_ready_for_aggregation_queues =
+            get_resource_by_name(table, "SecuredReadyForAggregationQueue".to_string())?;
+        let temp_storage = get_resource_by_name(table, "TempStorage".to_string())?;
+
+        Ok(Self {
+            secured_ready_for_aggregation_queues: create_clients_vec(
+                &secured_ready_for_aggregation_queues,
+                &client_options.queue_service,
+            ),
+            temp_storage: create_clients_vec(&temp_storage, &client_options.blob_service),
+        })
+    }
+}
+
+pub struct IngestClientResources {
+    client: KustoClient,
+    resources: Refreshing<Option<InnerIngestClientResources>>,
+    client_options: QueuedIngestClientOptions,
+}
+
+impl IngestClientResources {
+    pub fn new(client: KustoClient, client_options: QueuedIngestClientOptions) -> Self {
+        Self {
+            client,
+            resources: Arc::new(RwLock::new(Cached::new(None, RESOURCE_REFRESH_PERIOD))),
+            client_options,
+        }
+    }
+
+    /// Gets the latest resources from Kusto, updating the cached resources if they are expired
```

Comment:
comment: I think this comment is wrong - this function does not update the cached resources.  Perhaps just remove?